### PR TITLE
Admission Control Filter

### PR DESCRIPTION
**Pull request by tonya11en**
This is an implementation of success-rate based admission control via an HTTP filter. Based on the success rate, the filter will probabilistically reject outbound requests. Motivation and some background on this filter's calculations can be found in #9658.

The filter tracks request success rate for each worker thread and uses the information to reject requests with some probability dictated by the SR (in some rolling window) before forwarding to the upstream.This removes the need for any locks such as those found in the local ratelimit filter and the adaptive concurrency filter.

To avoid taking any locks, the success is calculated per-thread. We track the total request count and the total number of successes for use in the following rejection probability:

image

These values are accumulated each second and inserted into a deque that contains the per-second accumulated values for the entire rolling window. This allows us to efficiently phase out stale SR data as it is no longer in the sliding window.

Progress on #9658 and a resurfacing of the work in https://github.com/envoyproxy/envoy/pull/10230.

### CODE REVIEW

**Comment by tonya11en** [comment]
I'm having a hard time with the doc build, but the rest of the patch compiles and runs tests locally. Will figure this doc situation out as soon as I can get to it.

**Comment by mattklein123** [comment]
@tonya11en can you get this passing tests and I will take a look?

/wait

**Comment by mattklein123** [comment]
Window CI failure looks legit.

/wait

**Comment by tonya11en** [comment]
Windows CI passed. Removing wait.

**Comment by tonya11en** [comment]
@snowp it came up in conversation that you might be interested in taking a look at this. No expectation, just FYI.

File: api/envoy/extensions/filters/http/admission_control/v3alpha/admission_control.proto [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+// [#protodoc-title: Admission Control]
+// [#extension: envoy.filters.http.admission_control]
+
+message AdmissionControl {
+  // Default method of specifying what constitutes a failed request. All status codes that
+  // indicate a failed request must be explicitly specified if not relying on the default
+  // values.
+  message DefaultEvaluationCriteria {
+    // If HTTP statuses are unspecified, defaults to 5xx.
```
Comment (mattklein123) [inline api/envoy/extensions/filters/http/admission_control/v3alpha/admission_control.proto:31]: Can you comment on valid ranges here and make sure we have code checks for them?

File: api/envoy/extensions/filters/http/admission_control/v3alpha/admission_control.proto [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    repeated type.v3.Int32Range http_status = 1;
+
+    // GRPC status codes to consider as request successes. If unspecified, defaults to "OK".
```
Comment (mattklein123) [inline api/envoy/extensions/filters/http/admission_control/v3alpha/admission_control.proto:34]: Is there any link that we can have for the code to uint mapping?

File: api/envoy/extensions/filters/http/admission_control/v3alpha/admission_control.proto [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  // Defines how a request is considered a success/failure.
+  oneof evaluation_criteria {
+    option (validate.required) = true;
+
+    DefaultEvaluationCriteria default_eval_criteria = 2;
+  }
+
+  // The time window over which the success rate is calculated. The window is rounded to the nearest
+  // second. Defaults to 120s.
```
Comment (mattklein123) [inline api/envoy/extensions/filters/http/admission_control/v3alpha/admission_control.proto:50]: Can you clarify that this is a sliding window?

File: source/common/http/http2/codec_impl.cc [hunk 4]
```diff
@@ -317,8 +317,8 @@ void ConnectionImpl::StreamImpl::saveHeader(HeaderString&& name, HeaderString&&
 void ConnectionImpl::StreamImpl::submitTrailers(const HeaderMap& trailers) {
   std::vector<nghttp2_nv> final_headers;
   buildHeaders(final_headers, trailers);
-  int rc =
-      nghttp2_submit_trailer(parent_.session_, stream_id_, &final_headers[0], final_headers.size());
+  int rc = nghttp2_submit_trailer(parent_.session_, stream_id_, final_headers.data(),
```
Comment (mattklein123) [inline source/common/http/http2/codec_impl.cc:320]: merge master? Unrelated?

File: source/extensions/filters/http/admission_control/admission_control.h [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+class AdmissionControlFilterConfig {
+public:
+  AdmissionControlFilterConfig(const AdmissionControlProto& proto_config, Runtime::Loader& runtime,
+                               TimeSource& time_source, Runtime::RandomGenerator& random,
+                               Stats::Scope& scope, ThreadLocal::SlotPtr&& tls,
+                               std::unique_ptr<ResponseEvaluator> response_evaluator);
+  virtual ~AdmissionControlFilterConfig() = default;
+
+  virtual ThreadLocalController& getController() const {
+    return tls_->getTyped<ThreadLocalControllerImpl>();
```
Comment (mattklein123) [inline source/extensions/filters/http/admission_control/admission_control.h:58]: I don't remember if/when we discussed this, but what's the tradeoff for a global controller vs. per worker? Would it be better to go with global for better accuracy similar to what we do for adaptive concurrency?

File: source/extensions/filters/http/admission_control/admission_control.cc [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+double AdmissionControlFilterConfig::aggression() const {
+  return std::max<double>(1.0, aggression_ ? aggression_->value() : defaultAggression);
+}
+
+AdmissionControlFilter::AdmissionControlFilter(AdmissionControlFilterConfigSharedPtr config,
+                                               const std::string& stats_prefix)
+    : config_(std::move(config)), stats_(generateStats(config_->scope(), stats_prefix)) {}
+
+Http::FilterHeadersStatus AdmissionControlFilter::decodeHeaders(Http::RequestHeaderMap&, bool) {
+  if (!config_->filterEnabled() || decoder_callbacks_->streamInfo().healthCheck()) {
```
Comment (mattklein123) [inline source/extensions/filters/http/admission_control/admission_control.cc:55]: It makes sense to ignore health checks, but can you make sure this is documented in the final docs?

File: source/extensions/filters/http/admission_control/admission_control.cc [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    decoder_callbacks_->sendLocalReply(Http::Code::ServiceUnavailable, "", nullptr, absl::nullopt,
+                                       "denied by admission control");
+    stats_.rq_rejected_.inc();
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+
+  if (deferred_record_failure_ != nullptr) {
+    // To give a new value to the Cleanup() variable, we must cancel it first to avoid triggering
+    // the function it wraps.
+    deferred_record_failure_->cancel();
```
Comment (mattklein123) [inline source/extensions/filters/http/admission_control/admission_control.cc:69]: How can this ever happen?

File: source/extensions/filters/http/admission_control/admission_control.cc [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterTrailersStatus
+AdmissionControlFilter::encodeTrailers(Http::ResponseTrailerMap& trailers) {
+  if (expect_grpc_status_in_trailer_) {
+    absl::optional<GrpcStatus> grpc_status = Grpc::Common::getGrpcStatus(trailers, false);
+
+    // Status code must be sent in trailers.
+    ASSERT(grpc_status.has_value());
```
Comment (mattklein123) [inline source/extensions/filters/http/admission_control/admission_control.cc:112]: No guarantee the upstream does this. Don't use ASSERT here and handle errors (and test).

File: source/extensions/filters/http/admission_control/admission_control.cc [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+
+  if (deferred_record_failure_ != nullptr) {
+    // To give a new value to the Cleanup() variable, we must cancel it first to avoid triggering
+    // the function it wraps.
+    deferred_record_failure_->cancel();
+  }
+  deferred_record_failure_ =
+      std::make_unique<Cleanup>([this]() { config_->getController().recordFailure(); });
```
Comment (mattklein123) [inline source/extensions/filters/http/admission_control/admission_control.cc:72]: Perf nit: this really shouldn't need a heap allocation. Since I'm pretty sure there must not exist a cleanup already since this is per filter, can you just member allocate a Cleanup inline in the constructor?

File: source/extensions/filters/http/admission_control/config.cc [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+namespace HttpFilters {
+namespace AdmissionControl {
+
+static constexpr std::chrono::seconds defaultSamplingWindow{120};
+
+Http::FilterFactoryCb AdmissionControlFilterFactory::createFilterFactoryFromProtoTyped(
+    const envoy::extensions::filters::http::admission_control::v3alpha::AdmissionControl& config,
+    const std::string& stats_prefix, Server::Configuration::FactoryContext& context) {
+
+  std::string prefix = stats_prefix + "admission_control.";
```
Comment (mattklein123) [inline source/extensions/filters/http/admission_control/config.cc:24]: nit: const

File: source/extensions/filters/http/admission_control/config.cc [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        return std::make_shared<ThreadLocalControllerImpl>(context.timeSource(), sampling_window);
+      });
+
+  std::unique_ptr<ResponseEvaluator> response_evaluator;
+  switch (config.evaluation_criteria_case()) {
+  case AdmissionControlProto::EvaluationCriteriaCase::kDefaultEvalCriteria:
+    response_evaluator = std::make_unique<DefaultResponseEvaluator>(config.default_eval_criteria());
+    break;
+  case AdmissionControlProto::EvaluationCriteriaCase::EVALUATION_CRITERIA_NOT_SET:
+  default:
```
Comment (mattklein123) [inline source/extensions/filters/http/admission_control/config.cc:42]: I don't think you need a default?

File: source/extensions/filters/http/admission_control/admission_control.cc (hunk 7 above)
Comment (tonya11en) [reply source/extensions/filters/http/admission_control/admission_control.cc:69]: It's been quite a while since I wrote that piece, but I recall that it was happening in the unit tests. It couldn't happen in practice.

File: source/common/http/http2/codec_impl.cc (hunk 4 above)
Comment (tonya11en) [reply source/common/http/http2/codec_impl.cc:320]: This was failing in CI because of some scenario where `final_headers` had a size of 0, so it was causing undefined behavior. Calling `vector::data()` gives us the same information, but circumvents the undefined behavior in the empty vector case.

I have no idea how my change could have caused a regression here since I don't touch really touch anything unrelated to the new filter, so I just assumed it would be fine to do this since it's a better way of getting that pointer.

File: source/extensions/filters/http/admission_control/admission_control.cc (hunk 8 above)
Comment (tonya11en) [reply source/extensions/filters/http/admission_control/admission_control.cc:112]: My reason for this was that I read this part in the [gRPC over HTTP2 protocol](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md) (emphasis mine):

> Response-Headers & Trailers-Only are each delivered in a single HTTP2 HEADERS frame block. Most responses are expected to have both headers and trailers but Trailers-Only is permitted for calls that produce an immediate error. **Status must be sent in Trailers even if the status code is OK**.

I interpreted that to mean that we can always expect a status. This should definitely be handled differently, I'm just unsure of what to do in this case- maybe return a `INVALID_ARGUMENT` status?

File: source/extensions/filters/http/admission_control/admission_control.h (hunk 5 above)
Comment (tonya11en) [reply source/extensions/filters/http/admission_control/admission_control.h:58]: Adaptive concurrency required global control because of the way it took measurements. We needed high levels of accuracy because of the percentile calculations- we would be in danger of missing tail latencies. We also needed to globally limit the concurrency across all worker threads since it would no longer mean that limiting concurrency (on one thread) resulted in decreased load on the local host. It would almost be the same as using adaptive concurrency on the egress. There's also complications around synchronization of the measurement windows, etc. Global just made more sense.

For this filter, the measurements aren't as finicky and making it thread-local would simply just decrease the number of samples. This is fine, since there's not really a control loop that other threads can interfere with (not sure if I'm using that correctly, but I mean we're not making a change and then measuring its effect). The decision-making here can be decentralized and we still get roughly the same results.

The benefits to thread-local controllers is obviously that we don't need to take any locks on the deque/structures that store the sampling history. The drawbacks are that it's a bit more complicated to implement and we need a copy of the sampling history deque/structures for each thread. I decided that the memory overhead is minimal, so it's worth it to make it thread-local. In cases where memory utilization might raise an eyebrow, we'd be at high risk for lock contention.

File: source/extensions/filters/http/admission_control/config.cc (hunk 11 above)
Comment (tonya11en) [reply source/extensions/filters/http/admission_control/config.cc:42]: I don't, but this catches a scenario where a new evaluator is implemented and configured, but not handled here. Otherwise, we'd catch it by dereferencing a null pointer somewhere.

File: source/common/http/http2/codec_impl.cc (hunk 4 above)
Comment (mattklein123) [reply source/common/http/http2/codec_impl.cc:320]: Merge master. @antoniovicente already fixed this.

File: source/extensions/filters/http/admission_control/admission_control.h (hunk 5 above)
Comment (mattklein123) [reply source/extensions/filters/http/admission_control/admission_control.h:58]: That's fine. Add some comments.

File: source/extensions/filters/http/admission_control/admission_control.cc (hunk 7 above)
Comment (mattklein123) [reply source/extensions/filters/http/admission_control/admission_control.cc:69]: OK let's remove.

File: source/extensions/filters/http/admission_control/admission_control.cc (hunk 8 above)
Comment (mattklein123) [reply source/extensions/filters/http/admission_control/admission_control.cc:112]: I would synthesize a failure value and just increment it (or something like that). It doesn't really matter.

File: source/extensions/filters/http/admission_control/config.cc (hunk 11 above)
Comment (mattklein123) [reply source/extensions/filters/http/admission_control/config.cc:42]: You will only allow the compiler to catch it if you remove default, which is what you want.

File: source/extensions/filters/http/admission_control/admission_control.h (hunk 5 above)
Comment (snowp) [reply source/extensions/filters/http/admission_control/admission_control.h:58]: I think one potential drawback with TLS is that incoming traffic isn't necessarily evenly distributed between the workers, so it could result in cases where overall we're rejecting some % of traffic, but individual clients that are connected to workers that haven't seen much traffic might be able to pass through without any rejections.

You could imagine 3 clients: client 1 and 2 are connected to worker A and client 3 to worker B. Client 1 starts sending traffic thats failing, triggering throttling. Client 2 and 3 are both sending identical traffic, but only client 2 sees throttling because it happened to be handled by the worker that is also handling client 1.

That said I think TLS is fine assuming we document the behavior.

File: source/extensions/filters/http/admission_control/evaluators/default_evaluator.cc [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+  if (evaluation_criteria.http_status_size() > 0) {
+    for (const auto& range : evaluation_criteria.http_status()) {
+      http_success_fns_.emplace_back([range](uint64_t status) {
+        return (static_cast<uint64_t>(range.start()) <= status) &&
+               (status < static_cast<uint64_t>(range.end()));
+      });
+    }
+  } else {
+    // We default to all 5xx codes as request failures.
+    http_success_fns_.emplace_back([](uint64_t status) { return status < 500; });
```
Comment (snowp) [inline source/extensions/filters/http/admission_control/evaluators/default_evaluator.cc:27]: these two lines together got me real confused since the comment talks about what we consider request failures, then the next line defines the success condition instead. Maybe reword the comment?

File: source/extensions/filters/http/admission_control/evaluators/default_evaluator.cc [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  // GRPC status.
+  if (evaluation_criteria.grpc_status_size() > 0) {
+    for (const auto& status : evaluation_criteria.grpc_status()) {
+      grpc_success_codes_.emplace_back(status);
+    }
+  } else {
+    grpc_success_codes_ = decltype(grpc_success_codes_)({
```
Comment (snowp) [inline source/extensions/filters/http/admission_control/evaluators/default_evaluator.cc:36]: i would think you can elide the type here? afaik it would resolve to the `initializer_list` ctor

Comment (snowp) [inline source/extensions/filters/http/admission_control/evaluators/default_evaluator.cc:36]: docs say you default to Ok, this has much more than that

File: source/extensions/filters/http/admission_control/evaluators/default_evaluator.h [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+public:
+  DefaultResponseEvaluator(envoy::extensions::filters::http::admission_control::v3alpha::
+                               AdmissionControl::DefaultEvaluationCriteria evaluation_criteria);
+  // ResponseEvaluator
+  bool isHttpSuccess(uint64_t code) const override;
+  bool isGrpcSuccess(uint32_t status) const override;
+
+private:
+  std::vector<std::function<bool(uint64_t)>> http_success_fns_;
+  std::vector<uint64_t> grpc_success_codes_;
```
Comment (snowp) [inline source/extensions/filters/http/admission_control/evaluators/default_evaluator.h:25]: you probably want this to be `uint32_t`, both `enumToInt` and the extracted `status` is 32 bit

File: source/extensions/filters/http/admission_control/thread_local_controller.h [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+ * to update them, at the cost of decreasing the number of samples.
+ *
+ * The look-back window for request samples is accurate up to a hard-coded 1-second granularity.
+ * TODO (tonya11en): Allow the granularity to be configurable.
+ */
+class ThreadLocalControllerImpl : public ThreadLocalController,
+                                  public ThreadLocal::ThreadLocalObject {
+public:
+  ThreadLocalControllerImpl(TimeSource& time_source, std::chrono::seconds sampling_window);
+  ~ThreadLocalControllerImpl() override = default;
```
Comment (snowp) [inline source/extensions/filters/http/admission_control/thread_local_controller.h:47]: not necessary

File: source/extensions/filters/http/admission_control/thread_local_controller.cc [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+static constexpr std::chrono::seconds defaultHistoryGranularity{1};
+
+ThreadLocalControllerImpl::ThreadLocalControllerImpl(TimeSource& time_source,
+                                                     std::chrono::seconds sampling_window)
+    : time_source_(time_source), sampling_window_(sampling_window) {}
+
+void ThreadLocalControllerImpl::maybeUpdateHistoricalData() {
+  // Purge stale samples.
+  while (!historical_data_.empty() && ageOfOldestSample() >= sampling_window_) {
```
Comment (snowp) [inline source/extensions/filters/http/admission_control/thread_local_controller.cc:22]: if you want to same some cycles here i think you can compute the cutoff timestamp and compare each sample with that instead of computing the diff between sampling time and current time for each sample

might not matter, it depends on how many samples are ejected at a time

File: source/extensions/filters/http/admission_control/thread_local_controller.cc [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    removeOldestSample();
+  }
+
+  // It's possible we purged stale samples from the history and are left with nothing, so it's
+  // necessary to add an empty entry. We will also need to roll over into a new entry in the
+  // historical data if we've exceeded the time specified by the granularity.
```
Comment (snowp) [inline source/extensions/filters/http/admission_control/thread_local_controller.cc:28]: its not obvious to me how this granularity comes into play, maybe add some detail?

**Comment by tonya11en** [comment]
@mattklein123 requested this PR be split, so I'm closing this one out and submitting another that incorporates the current comments and omits the thread-local controller.
